In [3]:
from erddapy import ERDDAP
import netCDF4 as nc
import pandas as pd
from connec_functions import GDB
from collections import defaultdict

from accessibility import check_endpoint
from summarize import * #json_keys, data_frame

### EMSO ERIC ERDDAP server

In [53]:
# analysed endpoint:
endpoint_url = "https://erddap.emso.eu/erddap"

#make connection
emsoERDDAP = ERDDAP(server=endpoint_url, protocol="tabledap")

In [3]:
if check_endpoint(endpoint_url):
    print("The endpoint is machine-accessible.")
else:
    print("The endpoint is not machine-accessible.")

Checking endpoint: https://erddap.emso.eu/erddap
Endpoint is online: 200
Content type may not be machine-readable: text/html;charset=UTF-8
The endpoint is machine-accessible.


**Exploration of the allDatasets dataset**  
~ retrieving a dataset that lists all available datasets on the ERDDAP server. The returned DataFrame will contain metadata for each dataset available on that server  
~ essentially a catalog of all datasets hosted on the server, including essential metadata that allows you to identify and filter the datasets of interest  

In [54]:
# set dataset ID
emsoERDDAP.dataset_id = "allDatasets"

#Get data (as 2D dataframe)
EMSO_alldatasets_df = emsoERDDAP.to_pandas()

In [5]:
#explore columns
EMSO_alldatasets_df.columns

Index(['datasetID', 'accessible', 'institution', 'dataStructure',
       'cdm_data_type', 'class', 'title', 'minLongitude (degrees_east)',
       'maxLongitude (degrees_east)', 'longitudeSpacing (degrees_east)',
       'minLatitude (degrees_north)', 'maxLatitude (degrees_north)',
       'latitudeSpacing (degrees_north)', 'minAltitude (m)', 'maxAltitude (m)',
       'minTime (UTC)', 'maxTime (UTC)', 'timeSpacing (seconds)', 'griddap',
       'subset', 'tabledap', 'MakeAGraph', 'sos', 'wcs', 'wms', 'files',
       'fgdc', 'iso19115', 'metadata', 'sourceUrl', 'infoUrl', 'rss', 'email',
       'testOutOfDate', 'outOfDate', 'summary'],
      dtype='object')

In [6]:
# save
EMSO_alldatasets_df.to_csv("properties/EMSO_ERDDAP_overview_metadata.csv", index=False)
# view
EMSO_alldatasets_df

datasetID accessible  \
0                                 allDatasets     public   
1                        EMSO_OBSEA_CTD_30min     public   
2    EMSO_OBSEA_Besos_Buoy_Airmar_200WX_30min     public   
3     EMSO_OBSEA_Besos_Buoy_Airmar_200WX_full     public   
4                                 E2M3A_METEO     public   
..                                        ...        ...   
166                            W1M3A_deploy04     public   
167                            W1M3A_deploy05     public   
168                            W1M3A_deploy06     public   
169                            W1M3A_deploy07     public   
170                EMSO_OBSEA_AWAC_waves_full     public   

                                           institution dataStructure  \
0                                 Axiom Docker Install         table   
1                  Polytechnic University of Catalonia         table   
2                  Polytechnic University of Catalonia         table   
3                  Polytechnic University of Catalonia         table   
4    National Institute of Oceanography and Applied...         table   
..                                                 ...           ...   
166                 Consiglio Nazionale delle Ricerche         table   
167                 Consiglio Nazionale delle Ricerche         table   
168                 Consiglio Nazionale delle Ricerche         table   
169                 Consiglio Nazionale delle Ricerche         table   
170                Polytechnic University of Catalonia         table   

    cdm_data_type                    class  \
0           Other  EDDTableFromAllDatasets   
1           Point       EDDTableFromErddap   
2           Point       EDDTableFromErddap   
3           Point       EDDTableFromErddap   
4      TimeSeries       EDDTableFromErddap   
..            ...                      ...   
166    TimeSeries       EDDTableFromErddap   
167    TimeSeries       EDDTableFromErddap   
168    TimeSeries       EDDTableFromErddap   
169    TimeSeries       EDDTableFromErddap   
170         Point       EDDTableFromErddap   

                                                 title  \
0    * The List of All Active Datasets in this ERDD...   
1    CTD data at OBSEA Underwater Observatory 30 mi...   
2    Data from Airmar 200 WX weather station deploy...   
3    Data from Airmar 200 WX weather station deploy...   
4     E2M3A METEO timeSeries, NRT in situ Observations   
..                                                 ...   
166                         W1M3A data (201705-201806)   
167                         W1M3A data (202010-202107)   
168                         W1M3A data (202107-202204)   
169                         W1M3A data (202310-......)   
170    Wave data from an AWAC-AST 1 MHz ADCP at OBSEA    

     minLongitude (degrees_east)  maxLongitude (degrees_east)  \
0                            NaN                          NaN   
1                       1.752570                     1.752570   
2                       1.752570                     1.752570   
3                       1.752570                     1.752570   
4                      18.082417                    18.082417   
..                           ...                          ...   
166                     9.111700                     9.118163   
167                     9.111700                     9.118163   
168                     9.111700                     9.118163   
169                     9.106570                     9.135168   
170                     1.752570                     1.752570   

     longitudeSpacing (degrees_east)  ...  \
0                                NaN  ...   
1                                NaN  ...   
2                                NaN  ...   
3                                NaN  ...   
4                                NaN  ...   
..                               ...  ...   
166                              NaN  ...   
167                              NaN  ...   
168           

In [7]:
print(f"There are {len(EMSO_alldatasets_df['datasetID'].drop_duplicates())} datasets offered by the EMSO ERDDAP server")

There are 171 datasets offered by the EMSO ERDDAP server


In [8]:
#See if other metadata listed in allDatasets dataset
for line in EMSO_alldatasets_df.summary:
    print(line)

This dataset is a table which has a row of information for each dataset currently active in this ERDDAP.
CTD data measured at OBSEA underwater observatory
Weather station from  an Airmar 200WX deployed at OBSEA's Besos Buoy, NW mediterranean sea
Weather station from  an Airmar 200WX deployed at OBSEA's Besos Buoy, NW mediterranean sea full sensor data
E2M3A METEO timeSeries, NRT in situ Observations
E2M3A MRDT timeSeries, NRT in situ Observations
E2M3A PCO2PROA timeSeries, NRT in situ Observations
E2M3A PCO2PROW timeSeries, NRT in situ Observations
E2M3A SAMI timeSeries, NRT in situ Observations
E2M3A SBE16PLS timeSeries, NRT in situ Observations
E2M3A SBE37O timeSeries, NRT in situ Observations
E2M3A timeSeries ALL INSTRUMENTS, NRT in situ Observations
E2M3A timeSeries, data collected from 2021 to 2022
This dataset contains dissolved iron concentrations ((Fe(II) + Fe(III); \u00b5mol/l) acquired between September 2013 and July 2017 using the CHEMINI Fe, a CHEmical MINIaturized analyser

Exploration of **search information** for each data offered through the ERDDAP server

In [9]:
# get search information
searchinfo_df = pd.read_csv(emsoERDDAP.get_search_url(response="csv"))

In [10]:
searchinfo_df.columns

Index(['griddap', 'Subset', 'tabledap', 'Make A Graph', 'wms', 'files',
       'Title', 'Summary', 'FGDC', 'ISO 19115', 'Info', 'Background Info',
       'RSS', 'Email', 'Institution', 'Dataset ID'],
      dtype='object')

In [11]:
searchinfo_df

griddap                                             Subset  \
0        NaN  https://erddap.emso.eu/erddap/tabledap/allData...   
1        NaN  https://erddap.emso.eu/erddap/tabledap/EMSO_OB...   
2        NaN  https://erddap.emso.eu/erddap/tabledap/EMSO_OB...   
3        NaN  https://erddap.emso.eu/erddap/tabledap/EMSO_OB...   
4        NaN                                                NaN   
..       ...                                                ...   
166      NaN  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...   
167      NaN  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...   
168      NaN  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...   
169      NaN  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...   
170      NaN  https://erddap.emso.eu/erddap/tabledap/EMSO_OB...   

                                              tabledap  \
0    https://erddap.emso.eu/erddap/tabledap/allData...   
1    https://erddap.emso.eu/erddap/tabledap/EMSO_OB...   
2    https://erddap.emso.eu/erddap/tabledap/EMSO_OB...   
3    https://erddap.emso.eu/erddap/tabledap/EMSO_OB...   
4    https://erddap.emso.eu/erddap/tabledap/E2M3A_M...   
..                                                 ...   
166  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...   
167  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...   
168  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...   
169  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...   
170  https://erddap.emso.eu/erddap/tabledap/EMSO_OB...   

                                          Make A Graph  wms  files  \
0    https://erddap.emso.eu/erddap/tabledap/allData...  NaN    NaN   
1    https://erddap.emso.eu/erddap/tabledap/EMSO_OB...  NaN    NaN   
2    https://erddap.emso.eu/erddap/tabledap/EMSO_OB...  NaN    NaN   
3    https://erddap.emso.eu/erddap/tabledap/EMSO_OB...  NaN    NaN   
4    https://erddap.emso.eu/erddap/tabledap/E2M3A_M...  NaN    NaN   
..                                                 ...  ...    ...   
166  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...  NaN    NaN   
167  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...  NaN    NaN   
168  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...  NaN    NaN   
169  https://erddap.emso.eu/erddap/tabledap/W1M3A_d...  NaN    NaN   
170  https://erddap.emso.eu/erddap/tabledap/EMSO_OB...  NaN    NaN   

                                                 Title  \
0    * The List of All Active Datasets in this ERDD...   
1    CTD data at OBSEA Underwater Observatory 30 mi...   
2    Data from Airmar 200 WX weather station deploy...   
3    Data from Airmar 200 WX weather station deploy...   
4     E2M3A METEO timeSeries, NRT in situ Observations   
..                                                 ...   
166                         W1M3A data (201705-201806)   
167                         W1M3A data (202010-202107)   
168                         W1M3A data (202107-202204)   
169                         W1M3A data (202310-......)   
170    Wave data from an AWAC-AST 1 MHz ADCP at OBSEA    

                                               Summary  \
0    This dataset is a table which has a row of inf...   
1    CTD data measured at OBSEA underwater observat...   
2    Weather station from  an Airmar 200WX deployed...   
3    Weather station from  an Airmar 200WX deployed...   
4    E2M3A METEO timeSeries, NRT in situ Observatio...   
..                                                 ...   
166  Data from W1M3A observatory (062015-062016)\n\...   
167  Data from W1M3A observatory (062015-062016)\n\...   
168  Data from W1M3A observatory (062015-062016)\n\...   
169  Data from W1M3A observatory (062015-062016)\n\...   
170  Wave data collected and processed by an AWAC-A...   

                                                  FGDC  \
0                                                  NaN   
1    https://erddap.emso.eu/erddap/metadata/fgdc/xm...   
2    https://erddap.emso.eu/erddap/metadata/fgdc/xm...   
3    https://erddap.emso.eu/erddap/m

Exploration of **metadata information** available for each dataset offered through the ERDDAP server

In [55]:
# Metadata information for each dataset
metadatainfo_df = pd.DataFrame(columns=["Row Type", "Variable Name", "Attribute Name", "Data Type", "Value", "DatasetID"])
# get metadata for each dataset
for datasetID in EMSO_alldatasets_df['datasetID']:
    emsoERDDAP.dataset_id = datasetID
    try:
        _df = pd.read_csv(emsoERDDAP.get_info_url(response="csv")) #metadata retrieved via info_url
        _df["DatasetID"] = datasetID
        metadatainfo_df = pd.concat([metadatainfo_df, _df], ignore_index=True)
    except Exception as e:
        print(f"there was an error for {datasetID}: '{e}'")
        continue

# save to csv file
metadatainfo_df.to_csv("properties/EMSO_ERDDAP_dataset_metadata.csv", index=False)

metadatainfo_df

Row Type Variable Name Attribute Name Data Type  \
0      attribute     NC_GLOBAL  cdm_data_type    String   
1      attribute     NC_GLOBAL    Conventions    String   
2      attribute     NC_GLOBAL  creator_email    String   
3      attribute     NC_GLOBAL   creator_name    String   
4      attribute     NC_GLOBAL    creator_url    String   
...          ...           ...            ...       ...   
80364  attribute      depth_QC   actual_range      byte   
80365  attribute      depth_QC    conventions    String   
80366  attribute      depth_QC  flag_meanings    String   
80367  attribute      depth_QC    flag_values    String   
80368  attribute      depth_QC      long_name    String   

                                                   Value  \
0                                                  Other   
1                               COARDS, CF-1.6, ACDD-1.3   
2                                     nobody@example.com   
3                                   Axiom Docker Install   
4                          https://erddap.emso.eu/erddap   
...                                                  ...   
80364                                               7, 7   
80365                                OceanSITES QC Flags   
80366  unknown;good_data;probably_good_data;potential...   
80367                                    0;1;2;3;4;7;8;9   
80368        depth of measurements quality control flags   

                        DatasetID  
0                     allDatasets  
1                     allDatasets  
2                     allDatasets  
3                     allDatasets  
4                     allDatasets  
...                           ...  
80364  EMSO_OBSEA_AWAC_waves_full  
80365  EMSO_OBSEA_AWAC_waves_full  
80366  EMSO_OBSEA_AWAC_waves_full  
80367  EMSO_OBSEA_AWAC_waves_full  
80368  EMSO_OBSEA_AWAC_waves_full  

[80369 rows x 6 columns]

In [13]:
# explore returned metadata
metadatainfo_df.groupby(['DatasetID', 'Row Type']).nunique()

Variable Name  Attribute Name  \
DatasetID                      Row Type                                   
BB_505_ADCP                    attribute             17              71   
                               variable              16               0   
BB_567_SBE56                   attribute              8              72   
                               variable               7               0   
BB_584_SBE56                   attribute              8              72   
...                                                 ...             ...   
smartbay_obs_fluorometer_ecofl variable              12               0   
smartbay_obs_hour_mean         attribute             59             136   
                               variable              58               0   
smartbay_obs_pco2_contros      attribute             11             129   
                               variable              10               0   

                                          Data Type  Value  
DatasetID                      Row Type                     
BB_505_ADCP                    attribute          4    140  
                               variable           4      0  
BB_567_SBE56                   attribute          3    107  
                               variable           3      0  
BB_584_SBE56                   attribute          3    105  
...                                             ...    ...  
smartbay_obs_fluorometer_ecofl variable           4      0  
smartbay_obs_hour_mean         attribute          3    515  
                               variable           4      0  
smartbay_obs_pco2_contros      attribute          3    180  
                               variable           4      0  

[342 rows x 4 columns]

In [58]:
# Explore the dimensions of each dataset (Row Type = 'dimension')
dim_info = metadatainfo_df[ metadatainfo_df['Row Type'] == 'dimension'].groupby(['DatasetID', 'Variable Name']).agg({
        'Data Type': 'unique',
        'Value': 'unique'
    }).reset_index()

dim_info.to_csv("properties/EMSO_ERDDAP_dataset_metadata_dimensions.csv", index=False)

dim_info

Empty DataFrame
Columns: [DatasetID, Variable Name, Data Type, Value]
Index: []

In [15]:
# Explore the variables of each dataset (Row Type = 'variable')
#note: in netCDFs, attributes represent additional information 
var_info = metadatainfo_df[ metadatainfo_df['Row Type'] == 'variable'].groupby(['DatasetID', 'Variable Name']).agg({
        'Data Type': 'unique',
        'Value': 'unique'
    }).reset_index()

var_info.to_csv("properties/EMSO_ERDDAP_dataset__metadata_variables.csv", index=False)

var_info

DatasetID      Variable Name Data Type  Value
0                   BB_505_ADCP      CurrVelE_ADCP  [double]  [nan]
1                   BB_505_ADCP      CurrVelN_ADCP  [double]  [nan]
2                   BB_505_ADCP     CurrVelUp_ADCP  [double]  [nan]
3                   BB_505_ADCP         CurrVel_QC    [byte]  [nan]
4                   BB_505_ADCP        ECHO_BEAM_1  [double]  [nan]
...                         ...                ...       ...    ...
3922  smartbay_obs_pco2_contros     pco2_corrected   [float]  [nan]
3923  smartbay_obs_pco2_contros  pco2_corrected_qc     [int]  [nan]
3924  smartbay_obs_pco2_contros   site_bathy_depth  [double]  [nan]
3925  smartbay_obs_pco2_contros         station_id  [String]  [nan]
3926  smartbay_obs_pco2_contros               time  [double]  [nan]

[3927 rows x 4 columns]

In [16]:
# Explore the attributes of each dataset (Row Type = 'attribute')
metadatainfo_df[ metadatainfo_df['Row Type'] == 'attribute'].groupby(['DatasetID', 'Attribute Name']).nunique()

Row Type  Variable Name  \
DatasetID                 Attribute Name                                   
BB_505_ADCP               Conventions                   1              1   
                          Easternmost_Easting           1              1   
                          Northernmost_Northing         1              1   
                          Southernmost_Northing         1              1   
                          Westernmost_Easting           1              1   
...                                                   ...            ...   
smartbay_obs_pco2_contros update_interval               1              1   
                          valid_max                     1              3   
                          valid_min                     1              3   
                          wfd_waterbody_name            1              1   
                          wfd_waterbody_type            1              1   

                                                 Data Type  Value  
DatasetID                 Attribute Name                           
BB_505_ADCP               Conventions                    1      1  
                          Easternmost_Easting            1      1  
                          Northernmost_Northing          1      1  
                          Southernmost_Northing          1      1  
                          Westernmost_Easting            1      1  
...                                                    ...    ...  
smartbay_obs_pco2_contros update_interval                1      1  
                          valid_max                      1      3  
                          valid_min                      1      3  
                          wfd_waterbody_name             1      1  
                          wfd_waterbody_type             1      1  

[16435 rows x 4 columns]

In [17]:
# and the variables/dimensions with which they're associated 
attr_info = metadatainfo_df[ metadatainfo_df['Row Type'] == 'attribute'].groupby(['DatasetID', 'Attribute Name', 'Variable Name']).agg({
        'Data Type': 'unique',
        'Value': 'unique'
    }).reset_index()

attr_info.to_csv("properties/EMSO_ERDDAP_dataset_metadata_attributes.csv", index=False)

attr_info

DatasetID         Attribute Name Variable Name  \
0                    BB_505_ADCP            Conventions     NC_GLOBAL   
1                    BB_505_ADCP    Easternmost_Easting     NC_GLOBAL   
2                    BB_505_ADCP  Northernmost_Northing     NC_GLOBAL   
3                    BB_505_ADCP  Southernmost_Northing     NC_GLOBAL   
4                    BB_505_ADCP    Westernmost_Easting     NC_GLOBAL   
...                          ...                    ...           ...   
76437  smartbay_obs_pco2_contros              valid_min         depth   
76438  smartbay_obs_pco2_contros              valid_min      latitude   
76439  smartbay_obs_pco2_contros              valid_min     longitude   
76440  smartbay_obs_pco2_contros     wfd_waterbody_name     NC_GLOBAL   
76441  smartbay_obs_pco2_contros     wfd_waterbody_type     NC_GLOBAL   

      Data Type                                              Value  
0      [String]  [OceanSITES v1.4,SeaDataNet_1.0,COARDS,CF-1.6,...  
1      [double]                                         [17.19352]  
2      [double]                                          [41.3413]  
3      [double]                                          [41.3413]  
4      [double]                                         [17.19352]  
...         ...                                                ...  
76437  [double]                                              [0.0]  
76438  [double]                                            [-90.0]  
76439  [double]                                           [-180.0]  
76440  [String]                                 [Outer Galway Bay]  
76441  [String]                                          [Coastal]  

[76442 rows x 5 columns]

Exploration of **data** from each dataset offered through the ERDDAP server, by accessing as pd.DataFrame  
(less efficient than accessing as ncCF)

In [4]:
#list datasets
print(f"there are {len(EMSO_alldatasets_df['datasetID'].drop_duplicates())} datasets available via EMSO ERDDAP")

# Connect to server 
erddap = ERDDAP(server="https://erddap.emso.eu/erddap", protocol="tabledap")

summary_full_df = pd.DataFrame()
#first half (otherwise code takes too long to run)
for datasetID in EMSO_alldatasets_df['datasetID'][:len(EMSO_alldatasets_df) // 2]:
    print(f"Processing dataset: {datasetID}")
    
    try:
        # Set the dataset ID
        erddap.dataset_id = datasetID
        
        # Fetch data as a 2D dataframe with timeout handling
        dataset_df = erddap.to_pandas()
        
        # Process data        
        summ_df = data_frame(dataset_df, datasetID)
        summary_full_df = pd.concat([summary_full_df, summ_df])
        
    except Exception as e:
        print(f"There was an error for {datasetID}: {e}")

# write first half to csv file
summary_full_df.to_csv("properties/EMSO_ERDDAP_dataset_data_metadata.csv", index=False)
# view
summary_full_df
#Property ~ variables (attributes & dimensions not clear)

there are 171 datasets available via EMSO ERDDAP
Processing dataset: allDatasets
Processing dataset: EMSO_OBSEA_CTD_30min
Processing dataset: EMSO_OBSEA_Besos_Buoy_Airmar_200WX_30min
Processing dataset: EMSO_OBSEA_Besos_Buoy_Airmar_200WX_full
Processing dataset: E2M3A_METEO
Processing dataset: E2M3A_MRDT
Processing dataset: E2M3A_PCO2PROA
Processing dataset: E2M3A_PCO2PROW
Processing dataset: E2M3A_SAMI
Processing dataset: E2M3A_SBE16PLS
Processing dataset: E2M3A_SBE37O
Processing dataset: E2M3A_CTD_meteo_CO2_pH_NRT
Processing dataset: E2M3A_2021_2022_TS
Processing dataset: Emso_Azores_Chemini_IRON
Processing dataset: EMSO-AZORES_TCM3-1_2016-2017
Processing dataset: EMSO-AZORES_TCM3-1_2017-2018
Processing dataset: EMSO-AZORES_TCM3-1_2018-2019
Processing dataset: EMSO-AZORES_TCM3-1_2021-2022
Processing dataset: EMSO-AZORES_TCM3-2_2017-2018
Processing dataset: EMSO-AZORES_TCM3-2_2018-2019
Processing dataset: EMSO-AZORES_TCM3-2_2019-2020
Processing dataset: EMSO-AZORES_TCM3-2_2020-2021
Pr

DatasetID                  Property  Count  \
0                            allDatasets                 datasetID    171   
1                            allDatasets                accessible    171   
2                            allDatasets               institution    171   
3                            allDatasets             dataStructure    171   
4                            allDatasets             cdm_data_type    171   
..                                   ...                       ...    ...   
3   EMSO-AZORES_EGIM_Turbidity_2017-2018  longitude (degrees_east)  25469   
4   EMSO-AZORES_EGIM_Turbidity_2017-2018                 depth (m)  25469   
5   EMSO-AZORES_EGIM_Turbidity_2017-2018                   DEPH_QC  25469   
6   EMSO-AZORES_EGIM_Turbidity_2017-2018                TUR4 (NTU)  25469   
7   EMSO-AZORES_EGIM_Turbidity_2017-2018                   TUR4_QC  25469   

      Types               Example                UniqueValues  
0    object           allDatasets                          {}  
1    object                public                    {public}  
2    object  Axiom Docker Install                          {}  
3    object                 table                     {table}  
4    object                 Other  {Other, Point, TimeSeries}  
..      ...                   ...                         ...  
3   float64             -32.27562                 {-32.27562}  
4   float64                1700.0                    {1700.0}  
5     int64                     7                         {7}  
6   float64                 294.0                          {}  
7     int64                     1                         {1}  

[1438 rows x 6 columns]

In [6]:
#second half
for datasetID in EMSO_alldatasets_df['datasetID'][(len(EMSO_alldatasets_df) // 2):]:
    print(f"Processing dataset: {datasetID}")
    
    try:
        # Set the dataset ID
        erddap.dataset_id = datasetID
        
        # Fetch data as a 2D dataframe with timeout handling
        dataset_df = erddap.to_pandas()
        
        # Process data        
        summ_df = data_frame(dataset_df, datasetID)
        summary_full_df = pd.concat([summary_full_df, summ_df], ignore_index=True)
        #overwrite each time, try to get info from as much datasets as possible before error
        summary_full_df.to_csv("properties/EMSO_ERDDAP_dataset_data_metadata2.csv", index=False)
    except Exception as e:
        print(f"There was an error for {datasetID}: {e}")


summary_full_df.to_csv("properties/EMSO_ERDDAP_dataset_data_metadata2.csv", index=False)
summary_full_df
#note: doesn't list netCDF attributes

Processing dataset: EMSO-AZORES_CHEMINI_Total-Iron_2011-2012
Processing dataset: EMSO-AZORES_CHEMINI_Total-Iron_2013-2014
Processing dataset: EMSO-AZORES_CHEMINI_Total-Iron_2014-2015
Processing dataset: EMSO-AZORES_CHEMINI_Total-Iron_2016-2017
Processing dataset: EMSO-AZORES_CHEMINI_Total-Iron_2017-2018
Processing dataset: EMSO-AZORES_CHEMINI_Total-Iron_2018-2019
Processing dataset: EMSO-AZORES_CHEMINI_Total-Iron_2019-2020
Processing dataset: EMSO-AZORES_Seamon-East_Optode-O2_2013-2014
Processing dataset: EMSO-AZORES_Seamon-East_Optode-O2_2015-2016
Processing dataset: EMSO-AZORES_Seamon-East_Optode-O2_2016-2017
Processing dataset: EMSO-AZORES_Seamon-East_Optode-O2_2012-2013
Processing dataset: EMSO-AZORES_Wetlabs_Turbidity_2011-2012
Processing dataset: EMSO-AZORES_Wetlabs_Turbidity_2012-2013
Processing dataset: EMSO-AZORES_Wetlabs_Turbidity_2013-2014
Processing dataset: EMSO-AZORES_Wetlabs_Turbidity_2015-2016
Processing dataset: EMSO-AZORES_Wetlabs_Turbidity_2016-2017
Processing datase

: 

: 

View of some datasets

In [68]:
# set dataset ID
emsoERDDAP.dataset_id = "EMSO_OBSEA_CTD_30min"

#Get data (as 2D dataframe)
EMSO_EMSO_OBSEA_AWAC_waves_full_df = emsoERDDAP.to_pandas()
EMSO_EMSO_OBSEA_AWAC_waves_full_df

time (UTC)  latitude (degrees_north)  \
0       2022-01-01T00:00:00Z                  41.18212   
1       2022-01-01T00:30:00Z                  41.18212   
2       2022-01-01T01:00:00Z                  41.18212   
3       2022-01-01T01:30:00Z                  41.18212   
4       2022-01-01T02:00:00Z                  41.18212   
...                      ...                       ...   
240635  2022-12-31T21:30:00Z                  41.18212   
240636  2022-12-31T22:00:00Z                  41.18212   
240637  2022-12-31T22:30:00Z                  41.18212   
240638  2022-12-31T23:00:00Z                  41.18212   
240639  2022-12-31T23:30:00Z                  41.18212   

        longitude (degrees_east)  depth (m)      sensor_id  \
0                        1.75257       20.0  16P57353-6479   
1                        1.75257       20.0  16P57353-6479   
2                        1.75257       20.0  16P57353-6479   
3                        1.75257       20.0  16P57353-6479   
4                        1.75257       20.0  16P57353-6479   
...                          ...        ...            ...   
240635                   1.75257       20.0  16P57353-6479   
240636                   1.75257       20.0  16P57353-6479   
240637                   1.75257       20.0  16P57353-6479   
240638                   1.75257       20.0  16P57353-6479   
240639                   1.75257       20.0  16P57353-6479   

        PSAL (Dimensionless)  PRES (Decibars)  SVEL (Metres per second)  \
0                    34.0903           19.236                  1505.240   
1                    34.0896           19.241                  1505.250   
2                    34.0892           19.252                  1505.250   
3                    34.0891           19.275                  1505.250   
4                    34.0838           19.290                  1505.240   
...                      ...              ...                       ...   
240635               38.1734              NaN                  1507.588   
240636               38.1732              NaN                  1507.578   
240637               38.1730              NaN                  1507.557   
240638               38.1728              NaN                  1507.563   
240639               38.1729              NaN                  1507.578   

        CNDC (Siemens per metre)  TEMP (Degrees Celsius)  PSAL_QC  PRES_QC  \
0                        4.17190                 14.7835        3        1   
1                        4.17190                 14.7843        3        1   
2                        4.17196                 14.7854        3        1   
3                        4.17193                 14.7852        3        1   
4                        4.17141                 14.7857        3        1   
...                          ...                     ...      ...      ...   
240635                   4.63894                 15.1012        1        9   
240636                   4.63860                 15.0982        1        9   
240637                   4.63786                 15.0914        1        9   
240638                   4.63806                 15.0935        1        9   
240639                   4.63859                 15.0984        1        9   

        SVEL_QC  CNDC_QC  TEMP_QC  latitude_QC  longitude_QC  depth_QC  
0             1        1        1            7             7         7  
1             1        1        1            7             7         7  
2             1        1        1            7             7         7  
3             1        1        1            7             7         7  
4             1        1        1            7             7         7  
...         ...      ...      ...          ...           ...       ...  
240635        1        1        1            7             7         7  
240636        1        1        1            7             7         7  
240637        1        1        1            7             7         7  
240638  

In [69]:
#Get data (as ncCF)
EMSO_EMSO_OBSEA_AWAC_waves_full_nc = emsoERDDAP.to_ncCF()
EMSO_EMSO_OBSEA_AWAC_waves_full_nc

<class 'netCDF4._netCDF4.Dataset'>
root group (NETCDF3_CLASSIC data model, file format NETCDF3):
    cdm_data_type: Point
    Conventions: OceanSITES;EMSO, CF-1.6
    data_mode: R
    data_type: OceanSITES time-series data
    date_created: 2023-10-25T14:49:16Z
    date_modified: 2023-10-25T14:49:16Z
    Easternmost_Easting: 1.75257
    emso_facility: OBSEA
    featureType: Point
    format_version: 1.4
    geospatial_lat_max: 41.18212
    geospatial_lat_min: 41.18212
    geospatial_lat_units: degrees_north
    geospatial_lon_max: 1.75257
    geospatial_lon_min: 1.75257
    geospatial_lon_units: degrees_east
    geospatial_vertical_max: 20.0
    geospatial_vertical_min: 20.0
    geospatial_vertical_positive: down
    geospatial_vertical_units: m
    history: 2024-08-26T06:43:08Z (local files)
2024-08-26T06:43:08Z https://data.obsea.es/erddap/tabledap/EMSO_OBSEA_CTD_30min.ncCF
    id: EMSO_OBSEA_CTD_30min
    infoUrl: https://edmo.seadatanet.org/report/2150
    institution: Polytechnic 

#### Comparing the properties between ERDDAP servers (ARGO & EMSO ERDDAP servers)

Comparing metadata of each server's 'allDatasets' dataset

In [93]:
# Function to summarize DataFrame
def summarize_dataframe(df):
    summary = []
    for col in df.columns:
        col_summary = {
            'Column': col,
            'Data Type': df[col].dtype,
            'Number of Unique Values': df[col].nunique(),
            'Number of Missing Values': df[col].isna().sum(),
            'Sample Unique Values': df[col].unique()[:5]  # List first 5 unique values as sample
        }
        summary.append(col_summary)
    
    summary_df = pd.DataFrame(summary)
    return summary_df

In [98]:
# Load ARGO general metadata
ARGO_general_metadata = pd.read_csv("properties/ARGO_ERDDAP_overview_metadata.csv")
# Load EMSO general metadata
EMSO_general_metadata = pd.read_csv("properties/EMSO_ERDDAP_overview_metadata.csv")

In [102]:
if (ARGO_general_metadata.columns == EMSO_general_metadata.columns).all():
    print("'allDatasets' dataset of both ERDDAP servers contains the same metadata properties")
else:
    print("'allDatasets' dataset of both ERDDAP servers contains different metadata properties")

'allDatasets' dataset of both servers contain the same metadata properties


In [100]:
# Get summary
ARGO_summary_df = summarize_dataframe(ARGO_general_metadata)
ARGO_summary_df

Column Data Type  Number of Unique Values  \
0                         datasetID    object                       53   
1                        accessible    object                        2   
2                       institution    object                       12   
3                     dataStructure    object                        2   
4                     cdm_data_type    object                        7   
5                             class    object                        7   
6                             title    object                       52   
7       minLongitude (degrees_east)   float64                       23   
8       maxLongitude (degrees_east)   float64                       22   
9   longitudeSpacing (degrees_east)   float64                       10   
10      minLatitude (degrees_north)   float64                       22   
11      maxLatitude (degrees_north)   float64                       24   
12  latitudeSpacing (degrees_north)   float64                       10   
13                  minAltitude (m)   float64                        7   
14                  maxAltitude (m)   float64                       14   
15                    minTime (UTC)    object                       29   
16                    maxTime (UTC)    object                       37   
17            timeSpacing (seconds)   float64                       23   
18                          griddap    object                       35   
19                           subset    object                       13   
20                         tabledap    object                       17   
21                       MakeAGraph    object                       52   
22                              sos   float64                        0   
23                              wcs   float64                        0   
24                              wms    object                       32   
25                            files    object                        7   
26                             fgdc    object                       47   
27                         iso19115    object                       47   
28                         metadata    object                       52   
29                        sourceUrl    object                        3   
30                          infoUrl    object                       24   
31                              rss    object                       52   
32                            email    object                       52   
33                    testOutOfDate    object                        4   
34                        outOfDate   float64                        5   
35                          summary    object                       35   

    Number of Missing Values  \
0                          0   
1                          0   
2                          0   
3                          0   
4                          0   
5                          0   
6                          0   
7                          6   
8                          6   
9                         20   
10                         6   
11                         6   
12                        20   
13                        17   
14                        17   
15                         9   
16                         7   
17                        21   
18                        18   
19                        40   
20                        36   
21                         1   
22                        53   
23                        53   
24                        21   
25                        46   
26                         6   
27                         6   
28                         1   
29                         1   
30                         1   
31                         1   
32                         1   
33                        48   
34                        48   
35                         0   

                                 Sample Unique Values  
0   [allDatasets, OACP-Argo-Global, ArgoFloats, A

In [96]:
# Get summary
EMSO_summary_df = summarize_dataframe(EMSO_general_metadata)
EMSO_summary_df

Column Data Type  Number of Unique Values  \
0                         datasetID    object                      171   
1                        accessible    object                        1   
2                       institution    object                       21   
3                     dataStructure    object                        1   
4                     cdm_data_type    object                        3   
5                             class    object                        2   
6                             title    object                      136   
7       minLongitude (degrees_east)   float64                       59   
8       maxLongitude (degrees_east)   float64                       59   
9   longitudeSpacing (degrees_east)   float64                        0   
10      minLatitude (degrees_north)   float64                       57   
11      maxLatitude (degrees_north)   float64                       56   
12  latitudeSpacing (degrees_north)   float64                        0   
13                  minAltitude (m)   float64                       46   
14                  maxAltitude (m)   float64                       52   
15                    minTime (UTC)    object                      147   
16                    maxTime (UTC)    object                      149   
17            timeSpacing (seconds)   float64                        0   
18                          griddap   float64                        0   
19                           subset    object                      153   
20                         tabledap    object                      171   
21                       MakeAGraph    object                      171   
22                              sos   float64                        0   
23                              wcs   float64                        0   
24                              wms   float64                        0   
25                            files   float64                        0   
26                             fgdc    object                      170   
27                         iso19115    object                      170   
28                         metadata    object                      171   
29                        sourceUrl    object                        3   
30                          infoUrl    object                       25   
31                              rss    object                      171   
32                            email    object                      171   
33                    testOutOfDate    object                        4   
34                        outOfDate   float64                       10   
35                          summary    object                       90   

    Number of Missing Values  \
0                          0   
1                          0   
2                          0   
3                          0   
4                          0   
5                          0   
6                          0   
7                         12   
8                         12   
9                        171   
10                        12   
11                        12   
12                       171   
13                        28   
14                        28   
15                         8   
16                        16   
17                       171   
18                       171   
19                        18   
20                         0   
21                         0   
22                       171   
23                       171   
24                       171   
25                       171   
26                         1   
27                         1   
28                         0   
29                         0   
30                         0   
31                         0   
32                         0   
33                       161   
34                       161   
35                         0   

                                 Sample Unique Values  
0   [allDatasets, EMSO_OBSEA_CTD_30min, EMSO_OBSE

Comparing metadata of each server's datasets

In [6]:
# Load ARGO properties from ARGO ERDDAP server
ARGO_props = pd.read_csv("properties/ARGO_ERDDAP_dataset_metadata.csv")
# Load ARGO properties from EMSO ERDDAP server
EMSO_props = pd.read_csv("properties/EMSO_ERDDAP_dataset_metadata.csv")

Comparing Dimension properties

In [59]:
ARGO_dimensions = pd.read_csv("properties/ARGO_ERDDAP_dataset_metadata_dimensions.csv")
# alternative, same results:
#ARGO_dimensions = ARGO_props[ ARGO_props['Row Type'] == "dimension" ]

ARGO_dimensions = ARGO_dimensions.groupby('Variable Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
ARGO_dimensions

Variable Name  Count                                        Values_list
0          PRES      1                      [OS_DYFAMED_1994-2014_D_TSO2]
1         depth     29  [HF_75c7_5b60_95d8, HF_ac49_84ad_3eb6, SDC_BAL...
2         frame      1                              [traj_lonlat_init_t0]
3      latitude     33  [HF_75c7_5b60_95d8, HF_ac49_84ad_3eb6, OACP-Ar...
4     longitude     33  [HF_75c7_5b60_95d8, HF_ac49_84ad_3eb6, OACP-Ar...
5         ntraj      1                              [traj_lonlat_init_t0]
6          time     32  [HF_75c7_5b60_95d8, HF_ac49_84ad_3eb6, OS_DYFA...

In [60]:
EMSO_dimensions = pd.read_csv("properties/EMSO_ERDDAP_dataset_metadata_dimensions.csv")
# alternative, same results:
#EMSO_dimensions = EMSO_props[ EMSO_props['Row Type'] == "dimension" ]

EMSO_dimensions = EMSO_dimensions.groupby('Variable Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
EMSO_dimensions

Empty DataFrame
Columns: [Variable Name, Count, Values_list]
Index: []

In [46]:
# Dimension properties common between both ERDDAP servers
common_dimensions = ARGO_dimensions[ARGO_dimensions['Variable Name'].isin(EMSO_dimensions['Variable Name'])]
common_dimensions = common_dimensions['Variable Name'].drop_duplicates()
common_dimensions

Series([], Name: Variable Name, dtype: object)

In [47]:
# Dimension properties unique to ARGO ERDDAP server
dimensions_unique_to_ARGO = ARGO_dimensions[~ARGO_dimensions['Variable Name'].isin(EMSO_dimensions['Variable Name'])]
dimensions_unique_to_ARGO = dimensions_unique_to_ARGO['Variable Name'].drop_duplicates()
dimensions_unique_to_ARGO

0         PRES
1        depth
2        frame
3     latitude
4    longitude
5        ntraj
6         time
Name: Variable Name, dtype: object

In [48]:
# Dimension properties unique to EMSO ERDDAP server
dimensions_unique_to_EMSO = EMSO_dimensions[~EMSO_dimensions['Variable Name'].isin(ARGO_dimensions['Variable Name'])]
dimensions_unique_to_EMSO = dimensions_unique_to_EMSO['Variable Name'].drop_duplicates()
dimensions_unique_to_EMSO

Series([], Name: Variable Name, dtype: object)

In [61]:
# overview of compared Dimension properties
props_dim_summ = pd.DataFrame.from_dict({
    "ARGO_dim" : [len(ARGO_dimensions)],
    "EMSO_dim" : [len(EMSO_dimensions)],
    "common_dim" : [len(common_dimensions)],
    "ARGO_unique_dim" : [len(dimensions_unique_to_ARGO)],
    "EMSO_unique_dim" : [len(dimensions_unique_to_EMSO)]
})
props_dim_summ

ARGO_dim  EMSO_dim  common_dim  ARGO_unique_dim  EMSO_unique_dim
0         7         0           0                7                0

Comparing Variable properties

In [27]:
ARGO_variables = pd.read_csv("properties/ARGO_ERDDAP_dataset_metadata_variables.csv")
# alternative, same results:
#ARGO_variables = ARGO_props[ ARGO_props['Row Type'] == "variable" ]

ARGO_variables = ARGO_variables.groupby('Variable Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
ARGO_variables

Variable Name  Count                             Values_list
0     Access_ordering_of_data      1                    [SDC_BLS_DATA_TS_V1]
1          Access_restriction      1                    [SDC_BLS_DATA_TS_V1]
2     Alternative_cruise_name      1                    [SDC_BLS_DATA_TS_V1]
3    Alternative_station_name      1                    [SDC_BLS_DATA_TS_V1]
4                      BBP700      1          [OceanGlidersGDACTrajectories]
..                        ...    ...                                     ...
536  vertical_sampling_scheme      1                            [ArgoFloats]
537                        vn      2   [drifter_6hour_qc, drifter_hourly_qc]
538                       wcs      1                           [allDatasets]
539             wmo_inst_type      2  [ArgoFloats, ArgoFloats-synthetic-BGC]
540                       wms      1                           [allDatasets]

[541 rows x 3 columns]

In [28]:
EMSO_variables = pd.read_csv("properties/EMSO_ERDDAP_dataset_metadata_variables.csv")
# alternative, same results:
#EMSO_variables = EMSO_props[ EMSO_props['Row Type'] == "variable" ]

EMSO_variables = EMSO_variables.groupby('Variable Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
EMSO_variables

Variable Name  Count  \
0                  AIRT      2   
1               AIRT_QC      2   
2                  ALKY      2   
3               ALKY_QC      2   
4                  ATMP      1   
..                  ...    ...   
413     velocity_upward      2   
414  velocity_upward_qc      2   
415                 wcs      1   
416                 wms      1   
417              xyz_QC      1   

                                           Values_list  
0    [EMSO_OBSEA_Besos_Buoy_Airmar_200WX_30min, EMS...  
1    [EMSO_OBSEA_Besos_Buoy_Airmar_200WX_30min, EMS...  
2    [EMSO_OBSEA_Besos_Buoy_SA8065_30min, EMSO_OBSE...  
3    [EMSO_OBSEA_Besos_Buoy_SA8065_30min, EMSO_OBSE...  
4                         [Emso_Ligure_Dyfamed_FCO2TW]  
..                                                 ...  
413        [smartbay_obs_adcp, smartbay_obs_hour_mean]  
414        [smartbay_obs_adcp, smartbay_obs_hour_mean]  
415                                      [allDatasets]  
416                                      [allDatasets]  
417  [Emso_Western_Ligurian_Bathybot_Aquadopp_NetCD...  

[418 rows x 3 columns]

In [23]:
# Variable properties common between both ERDDAP servers
common_variables = ARGO_variables[ARGO_variables['Variable Name'].isin(EMSO_variables['Variable Name'])]
common_variables = common_variables['Variable Name'].drop_duplicates()
common_variables

19          latitude
20         longitude
52              time
62       institution
275          CSPD_QC
           ...      
820             OSAT
821          OSAT_QC
827             TUR4
828          TUR4_QC
832    platform_name
Name: Variable Name, Length: 67, dtype: object

In [24]:
# Variable properties unique to ARGO ERDDAP server
variables_unique_to_ARGO = ARGO_variables[~ARGO_variables['Variable Name'].isin(EMSO_variables['Variable Name'])]
variables_unique_to_ARGO = variables_unique_to_ARGO['Variable Name'].drop_duplicates()
variables_unique_to_ARGO

0                             chla
1                          chla_qc
2            config_mission_number
3                     cycle_number
4                      data_center
                  ...             
958    hist_stop_coast_TELUK_BENOA
959                traj_iU_init_t0
960                traj_jV_init_t0
961               traj_lat_init_t0
962               traj_lon_init_t0
Name: Variable Name, Length: 474, dtype: object

In [25]:
# Variable properties unique to EMSO ERDDAP server
variables_unique_to_EMSO = EMSO_variables[~EMSO_variables['Variable Name'].isin(ARGO_variables['Variable Name'])]
variables_unique_to_EMSO = variables_unique_to_EMSO['Variable Name'].drop_duplicates()
variables_unique_to_EMSO

0                    CurrVelE_ADCP
1                    CurrVelN_ADCP
2                   CurrVelUp_ADCP
3                       CurrVel_QC
4                      ECHO_BEAM_1
                   ...            
3890         sbe16plus_salinity_qc
3891       sbe16plus_soundvelocity
3892    sbe16plus_soundvelocity_qc
3893         sbe16plus_temperature
3894      sbe16plus_temperature_qc
Name: Variable Name, Length: 351, dtype: object

In [64]:
# overview of compared Variable properties
props_var_summ = pd.DataFrame.from_dict({
    "ARGO_var" : [len(ARGO_variables)],
    "EMSO_var" : [len(EMSO_variables)],
    "common_var" : [len(common_variables)],
    "ARGO_unique_var" : [len(variables_unique_to_ARGO)],
    "EMSO_unique_var" : [len(variables_unique_to_EMSO)]
})
props_var_summ

ARGO_var  EMSO_var  common_var  ARGO_unique_var  EMSO_unique_var
0       541       418          67              474              351

Comparing attribute properties

In [30]:
ARGO_attributes = pd.read_csv("properties/ARGO_ERDDAP_dataset_metadata_attributes.csv")
# alternative, same results:
#ARGO_attributes = ARGO_props[ ARGO_props['Row Type'] == "attribute" ]

#ARGO_attributes.head()
ARGO_attributes = ARGO_attributes.groupby('Attribute Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
ARGO_attributes

Attribute Name  Count  \
0       Acknowledgements      6   
1          Author_e_mail     23   
2                    CDI      4   
3                    CDO      4   
4               C_format     59   
..                   ...    ...   
309              version      2   
310        w_surf_option      1   
311  wmo_instrument_type      1   
312    wmo_platform_code      2   
313               zsigma      1   

                                           Values_list  
0    [SDC_NAT_CLIM_TS_V1_025_m, SDC_NAT_CLIM_TS_V1_...  
1    [SDC_BLS_CLIM_TS_V1_m, SDC_BLS_CLIM_TS_V1_s, S...  
2    [SDC_NAT_CLIM_TS_V1_025_m, SDC_NAT_CLIM_TS_V1_...  
3    [SDC_NAT_CLIM_TS_V1_025_m, SDC_NAT_CLIM_TS_V1_...  
4    [ArgoFloats, ArgoFloats, ArgoFloats, ArgoFloat...  
..                                                 ...  
309           [ArgoFloats-reference, OACP-Argo-Global]  
310                  [ariane_trajectories_qualitative]  
311                                   [copernicus-fos]  
312    [OS_DYFAMED_1994-2014_D_TSO2, drifter_6hour_qc]  
313                  [ariane_trajectories_qualitative]  

[314 rows x 3 columns]

In [31]:
EMSO_attributes = pd.read_csv("properties/EMSO_ERDDAP_dataset_metadata_attributes.csv")
# alternative, same results:
#EMSO_attributes = EMSO_props[ EMSO_props['Row Type'] == "attribute" ]

#EMSO_attributes.head()
EMSO_attributes = EMSO_attributes.groupby('Attribute Name').agg(
    Count=('DatasetID', 'size'),
    Values_list=('DatasetID', list)
).reset_index()
EMSO_attributes

Attribute Name  Count  \
0              Conventions    220   
1             DM_indicator     29   
2      Easternmost_Easting    159   
3    Northernmost_Northing    159   
4         Processing_level     34   
..                     ...    ...   
266              valid_min   1646   
267            valid_range     10   
268     wfd_waterbody_name      7   
269     wfd_waterbody_type      7   
270      wmo_platform_code    109   

                                           Values_list  
0    [BB_505_ADCP, BB_567_SBE56, BB_584_SBE56, BB_5...  
1    [Emso_Ligure_Dyfamed_SedimentTrap, Emso_Ligure...  
2    [BB_505_ADCP, BB_567_SBE56, BB_584_SBE56, BB_5...  
3    [BB_505_ADCP, BB_567_SBE56, BB_584_SBE56, BB_5...  
4    [Emso_Azores_Chemini_IRON, Emso_Azores_Chemini...  
..                                                 ...  
266  [E1M3A_20070528_20080226, E1M3A_20070528_20080...  
267  [Emso_Western_Ligurian_Bathybot_ADCP_NetCDF_20...  
268  [smartbay_obs_acoustic, smartbay_obs_adcp, sma...  
269  [smartbay_obs_acoustic, smartbay_obs_adcp, sma...  
270  [E1M3A_20070528_20080226, E1M3A_20080227_20090...  

[271 rows x 3 columns]

In [32]:
# Attribute properties common between both ERDDAP servers
common_attributes = ARGO_attributes[ARGO_attributes['Attribute Name'].isin(EMSO_attributes['Attribute Name'])]
common_attributes = common_attributes['Attribute Name'].drop_duplicates()
common_attributes

6                Conventions
13              DM_indicator
16       Easternmost_Easting
21     Northernmost_Northing
22              QC_procedure
               ...          
304                    units
305          update_interval
307                valid_max
308                valid_min
312        wmo_platform_code
Name: Attribute Name, Length: 135, dtype: object

In [33]:
# Attribute properties unique to ARGO ERDDAP server
attributes_unique_to_ARGO = ARGO_attributes[~ARGO_attributes['Attribute Name'].isin(EMSO_attributes['Attribute Name'])]
attributes_unique_to_ARGO = attributes_unique_to_ARGO['Attribute Name'].drop_duplicates()
attributes_unique_to_ARGO

0         Acknowledgements
1            Author_e_mail
2                      CDI
3                      CDO
4                 C_format
              ...         
306    user_manual_version
309                version
310          w_surf_option
311    wmo_instrument_type
313                 zsigma
Name: Attribute Name, Length: 179, dtype: object

In [34]:
# Attribute properties unique to EMSO ERDDAP server
attributes_unique_to_EMSO = EMSO_attributes[~EMSO_attributes['Attribute Name'].isin(ARGO_attributes['Attribute Name'])]
attributes_unique_to_EMSO = attributes_unique_to_EMSO['Attribute Name'].drop_duplicates()
attributes_unique_to_EMSO

4        Processing_level
5            QC_indicator
17     ancllary_variables
20      atlantos_EOV_name
21       atlantos_EOV_urn
              ...        
261      type_of_analysis
262           uncertainty
267           valid_range
268    wfd_waterbody_name
269    wfd_waterbody_type
Name: Attribute Name, Length: 136, dtype: object

In [63]:
# overview of compared Attribute properties
props_attr_summ = pd.DataFrame.from_dict({
    "ARGO_attr" : [len(ARGO_attributes)],
    "EMSO_attr" : [len(EMSO_attributes)],
    "common_attr" : [len(common_attributes)],
    "ARGO_unique_attr" : [len(attributes_unique_to_ARGO)],
    "EMSO_unique_attr" : [len(attributes_unique_to_EMSO)]
})
props_attr_summ

ARGO_attr  EMSO_attr  common_attr  ARGO_unique_attr  EMSO_unique_attr
0        314        271          135               179               136

In [65]:
props_var_summ

ARGO_var  EMSO_var  common_var  ARGO_unique_var  EMSO_unique_var
0       541       418          67              474              351

In [62]:
props_dim_summ

ARGO_dim  EMSO_dim  common_dim  ARGO_unique_dim  EMSO_unique_dim
0         7         0           0                7                0

### Analysis results:

- requires knowledge on ERDDAP servers & netCDF file format  
- ERDDAP server offers 170 datasets  
- mainly NetCDF data
- exploration of 'EMSO_ERDAP_dataset_metadata.csv' and 'EMSO_ERDDAP_dataset_data_metadata(2).csv', which shows information about the properties in each datasets (such as type, example, unique values, ...), shows:
    - in some cases, use of urls (e.g. for creator or OrcIDs)  
    - use of URLs of known, standard licenses 
    - opportunity for more use of persistent identifiers (e.g. station codes, ROR-id for institutes, ...)

#### Comparison of properties of data between the ARGO and EMSO ERDDAP servers

- the ERDDAP server interface return data in a semi-standard manner
- 'allDatasets' dataset returns information about the datasets that are available through both servers. The metadata to describe this information is the same in both servers.  
- both ERDDAP servers return netDCF datasets: data properties can be divided into three categories: dimensions, variables and attributes 
- the data properties of datasets have some degree of overlap
    - 67 of variable properties 
    - 135 of attribute properties 
- but there is still large number of properties unique to each server:
    - 7 of dimension properties
    - 825 of variable properties
    - 315 of attribute properties
- properties were analysed based on name only, no information available about the properties meaning/semantics 

